### ISSUE TITLE
Sorting by 'type' won't sort folders properly

### ISSUE THREAD

**Issue by Brunoesp** [ranger/ranger#1098]
#### Runtime Environment

- Operating system:  Arch Linux
- Terminal emulator: Urxvt
- Python version: 3.6.4 
- Ranger version: stable-1.9.0

#### Current Behavior
Folders won't sort properly under 'type' option. If initialized under type, folders aren't sorted, but if set to 'natural' then change to 'type' during runtime, folders keep sorted and files change as expected.

#### Expected Behavior
Setting rc.conf to sorting by 'type' should sort folders properly as files does.

**Comment by toonn** [comment]
Hmm, I can't reproduce this. ot keeps things sorted symbols before digits before caps before lower case.

**Comment by BrunoEspindola** [comment]
It's best reproducible if you open ranger, sort by mtime, for example, then change to type. The files will be sorted by extension but alphabetically will stay as mtime sorted then.

**Comment by Brunoesp** [comment]
It's best reproducible if you open ranger, sort by mtime, for example, then change to type. The files will be sorted by extension but alphabetically will stay as mtime sorted then.

**Comment by toonn** [comment]
Oh, sounds like what I was actually going to suggest as a solution. Have stable sorts, so if you sort alphabetically first then by type it would be sorted first by type then subsorted alphabetically.
I consider this a feature, not a bug. You can get what you want with 4 key presses, implementing what you want would cost us the ability to sort by type and subsort by whatever.
I'll leave this open for now in case someone else has something to say about it but unless there's a convincing argument I'll close this issue.

**Comment by Brunoesp** [comment]
I can't see any advantage of files sorted only by type by default. Anyway, is there a way to open ranger already sorted this way (type-natural)?

**Comment by toonn** [comment]
Only by type? That's not what happens for me. Every type is sorted by the order of the previes sort for me.

**Comment by Brunoesp** [comment]
I mean when I first open ranger and it reads from the rc.conf

**Comment by toonn** [comment]
You can add `chain set sort=natural; set sort=type` to the end of `rc.conf` I guess. You could also define a custom linemode and set that.

**Comment by Brunoesp** [comment]
It reads the first type and ignore the second.

**Comment by toonn** [comment]
Could you join the irc channel on [Freenode](https://webchat.freenode.net/)? Github's not well suited to realtime communication.

**Comment by Brunoesp** [comment]
Maybe switch label to bug? As discussed on irc, the commands behaviour should be the same on Ranger runtime as in the config file ('chain set sort=natural; set sort=type' works on the run but has no effect in the rc.conf).

**Comment by dlrn** [comment]
Yeah, I have the same problem, it's really annoying.

**Comment by toonn** [comment]
Well, I disagree it's a bug. Ranger's sorts don't stack (yet, hence the feature request label). The sorts are stable, so if you sort one way first then another you can manipulate the "subsort".

Feel free to open an issue about commands in `rc.conf` not having the same semantics as while running though, that would be a bug and I think we neglected to open one.

**Comment by Brunoesp** [comment]
You can sort of manipulate the "subsort" only in the current "buffer" (don't really know how it works). If you keep navigating in a deep file structure it falls back to the last issued sorting type.

**Comment by toonn** [comment]
Yep, that's how it works. The subsort is whatever the sort was there previously. What you're looking for is stacking stable sorts. Which requires some managing of such a stack to avoid heaping redundant sorts on there.

That's the feature request in this issue. @Vifon, this may actually fit the filterstack stuff? No managing but at least it'd make it possible manually, right?

**Comment by Brunoesp** [comment]
I think it would suffice at some level if at every folder entry the "onot" could be issued automatically. Is it even possible to automate this?

**Comment by Brunoesp** [comment]
> Well, I disagree it's a bug. Ranger's sorts don't stack (yet, hence the feature request label). The sorts are stable, so if you sort one way first then another you can manipulate the "subsort".
> 
> Feel free to open an issue about commands in `rc.conf` not having the same semantics as while running though, that would be a bug and I think we neglected to open one.

I don't think it's a semantics bug in rc.conf. What is really happening is that since the rc.conf file is read before the first folder buffer is opened, as I mentioned in the early comments, only the last issued sort is applied to the files.

**Comment by toonn** [comment]
Yes, `rc.conf` doesn't really need fixing for *this*, still a bit surprising sometimes though.

That's what I'm saying, we'd need to keep a stack of sorts to apply.

**Comment by Vifon** [comment]
@toonn, I don't think filterstack in the current form can handle it but it can be extended. It might have an optional `sort` method that would default to identity that would be called on the filtered file listing. Possibly after applying all filters for quicker processing. I'll add it to my TODO.

**Comment by cybarspace** [comment]
Is it possible to edit `rc.conf` so that the default behaviour is to sort directories alphabetically and other files by extension (not type, because I want jpg and png files in separate groups), with dotfiles after files and dot-directories after directories (when hidden files are shown)?

If it isn't possible yet, is there any plan to make a `sort_directories` option in `rc.conf`?

**Comment by toonn** [comment]
@cybarspace, your use case seems pretty unique so I doubt this'll be possible any time soon unless you implement it. Preferably as part of work to make sorting altogether more flexible.

**Comment by brunoesp** [comment]
I found a workaround for this issue. To sort naturally and then subsort by type (or in my case, extension), I added another line above the sort statement in the definition of sort in the _container/directory.py_ file:

`self.files_all.sort(key=sort_naturally_icase)`
`self.files_all.sort(key=sort_func)`

With this extra line of code and setting `set sort extension` in the _rc.conf_ file, subsorting works as expected.

**Comment by cybardev** [comment]
@brunoesp that works well, thanks a lot 😊

I adjusted it a little so it doesn't mess up other sorting algorithms:

```py
# sort directories naturally and sort files by extension
if (
    self.settings.sort_directories_first
    and self.settings.sort == "extension"
):
    self.files_all.sort(key=sort_naturally)

self.files_all.sort(key=sort_func)
```

It would be nice if we had more options other than just `files_all`, like `dirs_only` and `files_only`. For now, though, this seems to be a good-enough fix.
